flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/join/window/WindowJoinOperator.java

```diff
@@ -0,0 +1,558 @@
+                new WindowListState<>((InternalListState<RowData, Long, RowData>) leftListState);
+
+        ListStateDescriptor<RowData> rightRecordStateDesc =
+                new ListStateDescriptor<>(RIGHT_RECORDS_STATE_NAME, rightType);
+        ListState<RowData> rightListState =
+                getOrCreateKeyedState(windowSerializer, rightRecordStateDesc);
+        this.rightWindowState =
+                new WindowListState<>((InternalListState<RowData, Long, RowData>) rightListState);
+
+        // metrics
+        this.leftNumLateRecordsDropped = metrics.counter(LEFT_LATE_ELEMENTS_DROPPED_METRIC_NAME);
+        this.leftLateRecordsDroppedRate =
+                metrics.meter(
+                        LEFT_LATE_ELEMENTS_DROPPED_RATE_METRIC_NAME,
+                        new MeterView(leftNumLateRecordsDropped));
+        this.rightNumLateRecordsDropped = metrics.counter(RIGHT_LATE_ELEMENTS_DROPPED_METRIC_NAME);
+        this.rightLateRecordsDroppedRate =
+                metrics.meter(
+                        RIGHT_LATE_ELEMENTS_DROPPED_RATE_METRIC_NAME,
+                        new MeterView(rightNumLateRecordsDropped));
+        this.watermarkLatency =
+                metrics.gauge(
+                        WATERMARK_LATENCY_METRIC_NAME,
+                        () -> {
+                            long watermark = internalTimerService.currentWatermark();
+                            if (watermark < 0) {
+                                return 0L;
+                            } else {
+                                return internalTimerService.currentProcessingTime() - watermark;
+                            }
+                        });
+    }
+
+    @Override
+    public void close() throws Exception {
+        super.close();
+        collector = null;
+        functionsClosed = true;
+        if (joinCondition != null) {
+            joinCondition.close();
+        }
+    }
+
+    @Override
+    public void dispose() throws Exception {
+        super.dispose();
+        collector = null;
+        if (!functionsClosed) {
+            functionsClosed = true;
+            if (joinCondition != null) {
+                joinCondition.close();
+            }
+        }
+    }
+
+    @Override
+    public void processElement1(StreamRecord<RowData> element) throws Exception {
+        processElement(element, leftWindowEndIndex, leftLateRecordsDroppedRate, leftWindowState);
+    }
+
+    @Override
+    public void processElement2(StreamRecord<RowData> element) throws Exception {
+        processElement(element, rightWindowEndIndex, rightLateRecordsDroppedRate, rightWindowState);
+    }
+
+    private void processElement(
+            StreamRecord<RowData> element,
+            int windowEndIndex,
+            Meter lateRecordsDroppedRate,
+            WindowListState<Long> recordState)
+            throws Exception {
+        RowData inputRow = element.getValue();
+        long windowEnd = inputRow.getLong(windowEndIndex);
```

Comment:
Actually, this `windowEnd` is a timestamp without timezone field.
It can not just be compared to watermark.
You can take a look to `RecordsWindowBuffer`, which modify by https://github.com/apache/flink/pull/15485
You should add more cases for TIMESTAMP_LTZ supporting.